### PR TITLE
[opentitanlib] Add utility for programming and reading OTP parameters

### PR DESCRIPTION
**Pull request by jwnrt**
This is a WIP for programming and reading OTP parameters through opentitanlib.

Needed for the https://github.com/lowRISC/opentitan/issues/17390 and https://github.com/lowRISC/opentitan/issues/17391 manufacturing tests which need to provision tokens.

The docs for programming and reading the OTP are [here](https://opentitan.org/book/hw/ip/otp_ctrl/doc/programmers_guide.html#readout-sequence).

This PR includes a functest at `//sw/device/silicon_creator/manuf/lib:otp_ctrl_functest` which writes to and reads back the `MANUF_STATE` OTP parameter to test programming and reading.

### CODE REVIEW

**Comment by jwnrt** [comment]
CC @pamaury who I think needs this for another test.

File: sw/host/tests/manuf/otp_ctrl/BUILD [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        "@crate_index//:anyhow",
+        "@crate_index//:structopt",
+    ],
+)
+
+opentitan_functest(
+    name = "otp_ctrl_functest",
+    srcs = ["//sw/device/silicon_creator/manuf/lib:empty_functest.c"],
+    cw310 = cw310_params(
+        bitstream = "//hw/bitstream:rom_otp_dev",
```
Comment (timothytrippel) [inline sw/host/tests/manuf/otp_ctrl/BUILD:32]: once you move this functest to sw/device/silicon_creator/manuf/lib/BUILD, define a new `otp_image` similar to [this](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/manuf/lib/BUILD;drc=80627cccf544e11c8507176e781f52850f5f0551;l=323), e.g.:
```
otp_image(
    name = "otp_img_otp_ctrl_functest",
    src = "//hw/ip/otp_ctrl/data:otp_json_test_unlocked0",
    visibility = ["//visibility:private"],
)
```
and a new bitstream with:
```
bitstream_splice(
      name = "bitstream_otp_ctrl_functest",
      src = "//hw/bitstream:rom",
      data = "otp_img_otp_ctrl_functest{}"
      meminfo = "//hw/bitstream:otp_mmi",
      update_usr_access = True,
      visibility = ["//visibility:private"],
  )
```

That way the OTP image you have on the FPGA will be completely un-provisioned (i.e. all zeros), except for the LC partition which will encode the TEST_UNLOCKED0 state.

Comment (timothytrippel) [reply sw/host/tests/manuf/otp_ctrl/BUILD:32]: Note, if you do this, the ROM_EXEC_EN setting will also be 0, so you should have no problem connecting the JTAG, and you shouldn't need the idle_test.c, but it doesn't hurt to also put it in place.

File: sw/host/tests/manuf/otp_ctrl/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    )
+    .context("failed to read parameter EN_CSRNG_SW_APP_READ")?;
+
+    println!("EN_CSRNG_SW_APP_READ: {en_csrng_sw_app_read:#x?}");
+    //assert_eq!(en_csrng_sw_app_read, [0x01]);
+
+    // Program and then read back the `TEST_UNLOCK_TOKEN` which isn't pre-set.
+    let test_unlock_token = [0x89abcdef; 4];
+    OtpParam::write_param(&*jtag, DamParam::TestUnlockToken, &test_unlock_token)
+        .context("failed to write TEST_UNLOCK_TOKEN")?;
```
Comment (timothytrippel) [inline sw/host/tests/manuf/otp_ctrl/src/main.rs:64]: This shouldn't work as you have things configured, i.e., you are using the `//hw/bitstream:rom_otp_dev` bitstream, which uses the [`img_dev` otp image](https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/otp_ctrl/data/BUILD;drc=dcb3fec2d61a8f05ef4cc5b6fe2938693310ca65;l=458), which has the [STD_OTP_OVERLAYS](https://cs.opensource.google/opentitan/opentitan/+/master:rules/otp.bzl;drc=6d0bcbec6e7cf728436f13c181654b673b43a672;l=346), which includes a [set and locked SECRET0 partition](https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/otp_ctrl/data/BUILD;drc=dcb3fec2d61a8f05ef4cc5b6fe2938693310ca65;l=231).

If you define an empty OTP image + bitstream to use for this test like I describe above, this should work.

File: sw/host/opentitanlib/src/test_utils/otp_ctrl.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                    let addr = byte_addr + (idx * mem::size_of::<u64>()) as u32;
+                    Self::dmi_write(jtag, addr, access_granule, [data_words[0], data_words[1]])?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Perform a single read over the Direct Memory Interface.
```
Comment (timothytrippel) [inline sw/host/opentitanlib/src/test_utils/otp_ctrl.rs:107]: I believe its DAI = [Direct Access Interface](https://opentitan.org/book/hw/ip/otp_ctrl/doc/theory_of_operation.html#direct-access-interface-control)
```suggestion
    /// Perform a single read over the Direct Access Interface.
```

File: sw/host/opentitanlib/src/test_utils/otp_ctrl.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    ///
+    /// On success, returns an array of words `[lower, upper]` where `upper` is non-zero
+    /// for 64-bit granularity reads.
+    fn dmi_read(jtag: &dyn Jtag, byte_addr: u32, granule: Granularity) -> OtpParamResult<[u32; 2]> {
```
Comment (timothytrippel) [inline sw/host/opentitanlib/src/test_utils/otp_ctrl.rs:111]:
```suggestion
    fn dai_read(jtag: &dyn Jtag, byte_addr: u32, granule: Granularity) -> OtpParamResult<[u32; 2]> {
```

File: sw/host/opentitanlib/src/test_utils/otp_ctrl.rs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        if granule == Granularity::B64 {
+            jtag.read_memory32(dam_rdata1_addr, &mut rdata[1..=1])
+                .map_err(|source| OtpParamError::Jtag { source })?;
+        }
+
+        Ok(rdata)
+    }
+
+    /// Perform a single write over the Direct Memory Interface.
+    fn dmi_write(
```
Comment (timothytrippel) [inline sw/host/opentitanlib/src/test_utils/otp_ctrl.rs:143]: same here: DAI / Direct Access Interface

File: sw/host/opentitanlib/src/test_utils/otp_ctrl.rs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+            Ok(status.contains(OtpCtrlStatus::DAI_IDLE))
+        })
+        .map_err(|source| OtpParamError::WaitForIdle { source })
+    }
+}
+
+pub type OtpParamResult<T> = Result<T, OtpParamError>;
+
+/// Failures to write OTP parameters through Direct Access Memory.
```
Comment (timothytrippel) [inline sw/host/opentitanlib/src/test_utils/otp_ctrl.rs:202]:
```suggestion
/// Failures to write OTP parameters through Direct Access Interface.
```

**Review by timothytrippel — COMMENTED** [review]
Mostly LGTM, couple cleanups, then we can get this merged to unblock the remaining tests.

File: sw/host/opentitanlib/bindgen/BUILD [hunk 7]
```diff
@@ -43,6 +44,10 @@ rust_bindgen_library(
         "--allowlist-type=dif_lc_ctrl_token",
         "--allowlist-var=LC_CTRL_.*_BIT",
         "--allowlist-var=LC_CTRL_.*_REG_OFFSET",
+        "--allowlist-var=OTP_CTRL_.*_BIT",
+        "--allowlist-var=OTP_CTRL_.*_REG_OFFSET",
+        "--allowlist-var=OTP_CTRL_.*_OFFSET",
```
[the remark this answers is not in the document]
Comment (jwnrt) [reply sw/host/opentitanlib/bindgen/BUILD:49]: Sure, I was wondering if I should make it explicit we want both the `_REG_OFFSET` and `_OFFSET` (for DAI params) registers.

**Comment by jwnrt** [comment]
Thanks @timothytrippel, updated with your fixes and test passes now!

**Comment by jwnrt** [comment]
I've added the ability to read and calculate the digests of OTP partitions (locking them).

Included a case in the functest for these.

**Review by timothytrippel — APPROVED** [review]
LGTM, thanks @jwnrt

**Pull request merged**